File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.py
from itertools import product
def solution(users, emoticons):
    answer = [0, 0]
    data = [10, 20, 30, 40]
    
    discount = []
    def dfs(L,tmp):
        if L == len(tmp):
            discount.append(tmp[:])
            return
        else:
            for i in data:
                tmp[L]+=i
                dfs(L+1,tmp)
                tmp[L]-=i
    dfs(0,[0]*len(emoticons))
    
    for dis in discount:
        cnt = 0
        get = 0
        for u in users:
            pay = 0
            for j in range(len(dis)):
                if u[0] <= dis[j]:
                    pay += emoticons[j]*(100-dis[j])/100
            if pay >= u[1]:
                pay = 0
                cnt += 1
            get += pay
        if cnt >= answer[0]:
            if answer[0] == cnt:
                answer[1] = max(answer[1], get)
            else:
                answer[1] = get
            answer[0] = cnt
            
            
                
    
    return answer

File: 프로그래머스/2/150368. 이모티콘 할인행사/이모티콘 할인행사.java
import java.util.*;
class Solution {
    static int n;
    static int[] emoticons;
    static int[][] users;
    static int[] sales = {10,20,30,40};
    static int[] answer = new int[2];
    
    public int[] solution(int[][] users, int[] emoticons) {
        this.users = users;
        this.emoticons = emoticons;
        int[] rates = new int[emoticons.length];
        
        dfs(0,rates);    
        
        return answer;
    }
    public static void dfs(int L, int[] rates){
        if (L == rates.length ){
            calc(rates);
            return;
        }
        for(int i=0; i<sales.length;i++){
            rates[L] = sales[i];
            dfs(L+1, rates);
        }        
    }
    public static void calc(int[] rates){
        int[] ans = new int[2];
        for(int i=0; i<users.length;i++){
            int total = 0;
            for(int j=0; j<emoticons.length; j++){
                if(users[i][0]<=rates[j]){
                    total += emoticons[j]*(100-rates[j])/100;
                }
            }
            if(total >= users[i][1]){
                ans[0] ++;
            }else{
                ans[1]+=total;
            }
        }
        if(ans[0]>answer[0]){
            answer[0] = ans[0];
            answer[1] = ans[1];
        }
        if(ans[0] == answer[0] && ans[1] > answer[1]){
            answer[0] = ans[0];
            answer[1] = ans[1];
        }
        
    }
    
}